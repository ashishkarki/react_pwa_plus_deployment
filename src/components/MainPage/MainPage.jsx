import React, { useEffect } from 'react'

import './MainPage.css'
import CardList from '../CardList/CardList'
import MyHeader from '../MyHeader'
import SearchBox from '../SearchBox'
import Scroll from '../Scroll'
import ErrorBoundary from '../ErrorBoundary'

export const MainPage = ({
  searchTerm,
  requestRobots,
  onRequestRobots,
  onSearchTermChange,
}) => {
  const { isPending, robots } = requestRobots

  useEffect(() => {
    console.log('MainPage.js useEffect, disptaching action to fetch robots')
    onRequestRobots()
  }, [])

  const filteredRobots = robots.filter((robot) => {
    return robot.name.toLowerCase().includes(searchTerm.toLowerCase())
  })

  return isPending ? (
    <h1>Loading</h1>
  ) : (
    <div className="tc">
      <MyHeader />
      <SearchBox searchChange={onSearchTermChange} />
      <Scroll>
        <ErrorBoundary>
          <CardList id="cardList" robots={filteredRobots} />
        </ErrorBoundary>
      </Scroll>
    </div>
  )
}

export default MainPage
