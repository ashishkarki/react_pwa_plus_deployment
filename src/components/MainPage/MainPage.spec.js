import React from 'react'
import MainPage from './MainPage'

import renderer from 'react-test-renderer'

let component = null

beforeEach(() => {
  jest.clearAllMocks()

  const mockProps = {
    searchTerm: '',
    requestRobots: {
      isPending: false,
      robots: [],
    },
    onRequestRobots: jest.fn(),
    onSearchTermChange: jest.fn(),
  }

  component = renderer.create(<MainPage {...mockProps} />)
})

it('renders without crashing', () => {
  expect(component.toJSON()).toMatchSnapshot()
})

it('calls onSearchTermChange when input changes', () => {
  const mockEvent = {
    target: {
      value: 'test',
    },
  }

  const rootInstance = component.root
  const input = rootInstance.findByType('input')
  input.props.onChange(mockEvent)

  expect.assertions(1)
  expect(component.toJSON()).toMatchSnapshot()
})

it('filters robots correctly for empty case', () => {
  const rootInstance = component.root
  const cardListComponent = rootInstance.findByProps({ id: 'cardList' })
  const cardListFilteredRobots = cardListComponent.props.robots

  expect(cardListFilteredRobots).toEqual([])
})

it('filters robots correctly for finding something case', () => {
  const testRobot = {
    id: 1,
    name: 'test',
    email: 'a@email.com',
  }
  const mockProps = {
    searchTerm: 'es',
    requestRobots: {
      isPending: false,
      robots: [testRobot],
    },
    onRequestRobots: jest.fn(),
    onSearchTermChange: jest.fn(),
  }
  component = renderer.create(<MainPage {...mockProps} />)

  const rootInstance = component.root
  const cardListComponent = rootInstance.findByProps({ id: 'cardList' })
  const cardListFilteredRobots = cardListComponent.props.robots

  expect(cardListFilteredRobots).toEqual([testRobot])
})
