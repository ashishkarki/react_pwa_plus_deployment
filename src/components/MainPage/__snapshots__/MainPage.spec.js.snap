// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls onSearchTermChange when input changes 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    RoboFriends
  </h1>
  <div
    className="css-1u4s5v3 pa2"
  >
    <input
      aria-label="Search Robots"
      className="pa3 ba b--green bg-lightest-blue"
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "target": Object {
                  "value": "test",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      placeholder="search robots"
      type="search"
    />
  </div>
  <div
    style={
      Object {
        "border": "5px solid black",
        "height": "800px",
        "overflow": "scroll",
      }
    }
  >
    <div
      className="css-mnwtuh"
    />
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    RoboFriends
  </h1>
  <div
    className="css-1u4s5v3 pa2"
  >
    <input
      aria-label="Search Robots"
      className="pa3 ba b--green bg-lightest-blue"
      onChange={[MockFunction]}
      placeholder="search robots"
      type="search"
    />
  </div>
  <div
    style={
      Object {
        "border": "5px solid black",
        "height": "800px",
        "overflow": "scroll",
      }
    }
  >
    <div
      className="css-mnwtuh"
    />
  </div>
</div>
`;
